evm/src/recursive_verifier.rs

```diff
@@ -328,6 +567,31 @@ pub fn add_virtual_all_proof<F: RichField + Extendable<D>, const D: usize>(
     }
 }
 
+pub fn add_virtual_recursive_all_proof<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    verifier_data: &[VerifierCircuitData<F, C, D>; NUM_TABLES],
+) -> RecursiveAllProofTargetWithData<D> {
+    let recursive_proofs = std::array::from_fn(|i| {
+        let verifier_data = &verifier_data[i];
+        builder.add_virtual_proof_with_pis(&verifier_data.common)
+    });
+    let verifier_data = std::array::from_fn(|i| {
+        let verifier_data = &verifier_data[i];
+        VerifierCircuitTarget {
+            constants_sigmas_cap: builder
```

Comment:
Is `constants_sigmas_cap` currently unconstrained? If so it seems like a security issue; I think the VK should be baked into the circuit, i.e. set to some constants. Maybe like this?

```rust
constants_sigmas_cap: builder.constant_merkle_cap(verifier_data.verifier_only.constants_sigmas_cap)
```

(`constant_merkle_cap` doesn't exist yet)

Might be a big odd to have in this function though, since this part wouldn't be virtual.

Reply:
Oh yeah of course. I implemented your suggestion and added a comment to the function saying the verifier data is not virtual.